fil_fungible_token/src/token/state.rs

```diff
@@ -494,33 +373,104 @@ mod test {
         ActorID,
     };
 
-    use crate::blockstore::MemoryBlockstore;
-
     use super::TokenState;
+    use crate::blockstore::SharedMemoryBlockstore;
 
     #[test]
     fn it_instantiates() {
-        let bs = MemoryBlockstore::new();
-        let state = TokenState::new(&bs).unwrap();
-        let cid = state.save(&bs).unwrap();
-        let saved_state = TokenState::load(&bs, &cid).unwrap();
+        let bs = &SharedMemoryBlockstore::new();
+        let state = TokenState::new(bs).unwrap();
+        let cid = state.save(bs).unwrap();
+        let saved_state = TokenState::load(bs, &cid).unwrap();
         assert_eq!(state, saved_state);
     }
 
     #[test]
-    fn it_increases_balance_of_new_actor() {
-        let bs = MemoryBlockstore::new();
-        let mut state = TokenState::new(&bs).unwrap();
+    fn it_increases_balance_from_zero() {
+        let bs = &SharedMemoryBlockstore::new();
+        let mut state = TokenState::new(bs).unwrap();
         let actor: ActorID = 1;
 
         // Initially any actor has an implicit balance of 0
-        assert_eq!(state.get_balance(&bs, actor).unwrap(), BigInt::zero());
+        assert_eq!(state.get_balance(bs, actor).unwrap(), BigInt::zero());
 
         let amount = BigInt::from(100);
-        state.increase_balance(&bs, actor, &amount).unwrap();
-        let new_cid = state.save(&bs).unwrap();
+        state.change_balance_by(bs, actor, &amount).unwrap();
+
+        assert_eq!(state.get_balance(bs, actor).unwrap(), amount);
+    }
+
+    #[test]
+    fn it_fails_to_decrease_balance_below_zero() {
+        let bs = &SharedMemoryBlockstore::new();
+        let mut state = TokenState::new(bs).unwrap();
+        let actor: ActorID = 1;
+
+        // can't decrease from zero
+        let err = state
+            .change_balance_by(bs, actor, &BigInt::from(-1))
+            .unwrap_err();
+        assert_eq!(
+            err.to_string(),
+            "resulting balance was negative adding -1 to 0"
+        );
```

Comment:
testing against the error msg here seems a little imprecise, probably indicating the misuse or innappropriateness of `anyhow` within this library.

Reply:
Probably better off using `thiserror` to define more concrete error types we can match against. In this case especially it could mean having a nice "insufficient balance" error or something which included the values.

Both crates suggest using `thiserror` for library-style code:
https://github.com/dtolnay/thiserror#comparison-to-anyhow
https://github.com/dtolnay/anyhow#comparison-to-thiserror